lib/clouds/libcloud_common.py

```diff
@@ -886,89 +981,114 @@ def vmcreate(self, obj_attr_list) :
 
             obj_attr_list["last_known_state"] = "about to send create request"
 
-            _image = self.get_images(obj_attr_list)
-            self.get_networks(obj_attr_list)
+            _mark_a = time()
+            obj_attr_list["libcloud_image_inst"] = self.get_images(obj_attr_list)
+            self.annotate_time_breakdown(obj_attr_list, "get_imageid_time", _mark_a)
 
             obj_attr_list["config_drive"] = False
 
             _status, _fmsg = self.vvcreate(obj_attr_list, LibcloudCmds.catalogs.cbtool[_credentials_list])
 
+            if "userdata" in obj_attr_list and str(obj_attr_list["userdata"]).lower() != "false" :                
+                obj_attr_list["userdata"] = self.populate_cloudconfig(obj_attr_list)
+                obj_attr_list["config_drive"] = True
+            else :
+                obj_attr_list["config_drive"] = False
+                obj_attr_list["userdata"] = None
+                
             self.common_messages("VM", obj_attr_list, "creating", 0, '')
 
-            if self.use_ssh_keys :
-
-                keys = []
-    
-                tmp_keys = obj_attr_list["key_name"].split(",")
-                for dontcare in range(0, 2) :
-                    for tmp_key in tmp_keys :
-                        for key in LibcloudCmds.keys[_credentials_list] :
-                            if tmp_key in [key.name, key.extra["id"]] and key.extra["id"] not in keys and key.name not in keys :
-                                keys.append(key.extra["id"])
-    
-                    if len(keys) >= len(tmp_keys) :
-                        break
-    
-                    cbdebug("Only found " + str(len(keys)) + " keys. Refreshing key list...", True)
-                    LibcloudCmds.keys[_credentials_list] = LibcloudCmds.catalogs.cbtool[_credentials_list].list_key_pairs()
-
-                extra["ssh_keys"] = keys
-
-                if len(keys) != len(tmp_keys) :
-                    raise CldOpsException("Not all SSH keys exist. Check your configuration: " + obj_attr_list["key_name"], _status, True)
+            extra = deepcopy(self.extra)
 
-            # Currently, regions and VMCs are the same in libcloud based adapters
-            obj_attr_list["region"] = _region = obj_attr_list["vmc_name"]
+            _location = self.get_region_from_vmc_name(obj_attr_list)
+            
             extra.update(self.pre_vmcreate(obj_attr_list, extra))
 
-            if self.use_locations :
-                _location = [x for x in LibcloudCmds.locations if x.id == obj_attr_list["region"]][0]
-            else :
-                _location = False 
-
+            obj_attr_list["libcloud_size_inst"] = False
```

Comment:
Please make use of pre_vmcreate. This is were modifications to the dictionary should happen, via callsbacks, as opposed to additionally stubbed functions. If the callback doesn't exist, then it just becomes a no-op in the cloud's adapter without needing to define the stub at all.

Reply:
Please see my comment on `self.vmcreate_kwargs["ex_custom_data"] = obj_attr_list["userdata"].` (i.e., `vmcreate_kwargs` requires information that is only available after `pre_vmcreate`).